Inconsistent restore results based on restore order

### NuGet Product Used

MSBuild.exe, NuGet.exe, NuGet SDK

### Product Version

nuget 6.9.1.3, net6

### Worked before?

_No response_

### Impact

I'm unable to use this version

### Repro Steps & Context

# Context
We use nuget packages for our native libraries but I think this may apply to any kind of package.

Our packages are built as follows:

A package containing binary files, one file `CopyBinaries.targets` to copy these files to output using `ReferenceCopyLocalPaths` and files in `build` and `buildTransitive` importing `CopyBinaries.targets` each.
```
.
└── MyPackage-binaries/
    ├── build/
    │   └── MyPackage-binaries.targets
    ├── buildCommon/
    │   └── CopyBinaries.targets
    ├── buildTransitive/
    │   └── MyPackage-binaries.targets
    ├── nativecontent/
    │   └── bin/
    │       └── MyPackage.dll
    └── MyPackage-binaries.nuspec
```

A package containing lib/header files, and a targets file setting `AdditionalLibraryDirectories`, `AdditionalIncludeDirectories` etc.
This package has the package above as a dependency.
```
Lib/header package:
.
└── MyPackage/
    ├── build/
    │   └── native/
    │       └── MyPackage.targets
    ├── nativecontent/
    │   ├── lib/
    │   │   └── MyPackage.lib
    │   └── include/
    │       └── MyPackage.h
    └── MyPackage.nuspec
```

We have project `A` which is a Net6 C++/CLI project using `PackageReference` to consume `MyPackage` directly and `MyPackage-binaries` through the transitive dependency. This works fine if `A` is not referenced anywhere itself.

However, when we add a managed Net6 project `B` to the solution referencing `A` the result is different and dependent on the order the projects are restored. If `A` is restored first targets files from both packages are imported correctly. But if `B` is restored first project `A` only consumes the transitive package and not the directly referenced one. As the restore order can change the outcome is unpredictable and `A` may not compile when it otherwise does.

# Expected Behavior
The generated `A.vcxproj.nuget.g.targets` imports `MyPackage\build\native\MyPackage.targets` and `MyPackage-binaries\buildTransitive\MyPackage-binaries.targets` regardless of restore order.

# Actual Behavior
If the transitive project `B` is restored first the project `A` will only consume the transitive package.

# Workaround
Add a `MyPackage\buildTransitive\MyPackage.targets` file mimicking the `MyPackage\build\native\MyPackage.targets` file. This would import vcxproj related tags into unrelated transitive projects however.

# Reproduction Steps
Using the archive below run a clean on the project `NativeCore` to make sure the `nuget.g.targets` file will be regenerated and then call either `nuget.exe restore` from CLI or rebuild that project from within VS/MsBuild. Most of the times `NativeCore` is restored first but when not the build fails as the header can't be found.

[NugetIssueReportData.zip](https://github.com/NuGet/Home/files/14619517/NugetIssueReportData.zip)
The file contains three nuget packages:
- `NativeTestPackage`: Include a header `test_include.h` into a vcxproj and transitively consume binaries through the next package
- `NativeTestPackage-binaries`: Copy a dummy file `thiswouldbeadll.txt` to the output
- `NativeTestPackageWithTransitive`: Same as `NativeTestPackage` but with a `buildTransitive` folder showing the workaround.

It also contains a VS solution with the following projects:

- `NativeCore`: C++/CLI project targeting net6 and referencing `NativeTestPackage`. Change this to `NativeTestPackageWithTransitive` for the workaround
- `NetCoreTransitive` referencing the former

# Analysis

I just compiled and debugged into a custom build of nuget.exe briefly but it appears when assets are resolved through a transitive project first the `LibraryIncludeFlags` do not contain `LibraryIncludeFlags.Build`. As assets are only resolved once before restoring all projects a following projects would find the asset to have been listed with no native targets files to import.

### Verbose Logs

_No response_

Hi @nkolev92.

Thanks for the reply.

> When you say restore order, how do you determine that? Restore also parallelizes projects, so often times they get completed around the same time.

I based this on the output in the VS/terminal console.

```text
Build started at 03:17...
1>------ Build started: Project: NativeCore, Configuration: Debug x64 ------
Restored C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\NativeCore\NativeCore.vcxproj (in 12 ms).
Restored C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\NetCoreTransitive\NetCoreTransitive.csproj (in 26 ms).
1>ManagedClass.cpp
1>.NETCoreApp,Version=v6.0.AssemblyAttributes.cpp
1>Generating Code...
1>NativeCoreNugetTest.cpp
1>ManagedClass.obj : MSIL module encountered; incremental linking is disabled for MSIL; performing full link
1>NativeCore.vcxproj -> C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\bin\Debug\NativeCore\NativeCore.dll
2>------ Build started: Project: NetCoreTransitive, Configuration: Debug x64 ------
2>NetCoreTransitive -> C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\bin\Debug\NetCoreTransitive\x64\Debug\NetCoreTransitive.dll
========== Build: 2 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 03:18 and took 03,984 seconds ==========
```

Every time I see this order the `NativeCore.vcxproj.nuget.g.targets` file contains two package references and the project compiles.

```text
Build started at 03:19...
1>------ Build started: Project: NativeCore, Configuration: Debug x64 ------
Restored C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\NetCoreTransitive\NetCoreTransitive.csproj (in 11 ms).
Restored C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\NativeCore\NativeCore.vcxproj (in 17 ms).
1>ManagedClass.cpp
1>.NETCoreApp,Version=v6.0.AssemblyAttributes.cpp
1>Generating Code...
1>NativeCoreNugetTest.cpp
1>C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\NativeCore\NativeCoreNugetTest.cpp(2,10): error C1083: Cannot open include file: 'test_include.h': No such file or directory
1>Done building project "NativeCore.vcxproj" -- FAILED.
2>------ Build started: Project: NetCoreTransitive, Configuration: Debug x64 ------
2>CSC : error CS0006: Metadata file 'C:\Users\bengt\Desktop\NugetIssueReportData\NugetTest\bin\Debug\NativeCore\NativeCore.dll' could not be found
2>Done building project "NetCoreTransitive.csproj" -- FAILED.
========== Build: 0 succeeded, 2 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 03:19 and took 01,781 seconds ==========

```

And every time this order happens the `NativeCore.vcxproj.nuget.g.targets` contains only one package and the header from the package is not found.

I triggered these above just by `Build->Clean Solution` and `Build->Build Solution` in VS.

> I really do appreciate all the work you've gone through, but if you can provide a self contained repro with the packages as nupkgs, rather than the nuspec, it'd make it even easier.

I wasn't sure if you guys are comfortable using third-party packages. Here's a new repro with packages prebuilt in the `NativeTestPackage\package` folder and the nuget.config pointing to that directory as a source. Let me know if this still doesn't work for you.
[NugetIssueReportData2.zip](https://github.com/NuGet/Home/files/14830241/NugetIssueReportData2.zip)

> Can you clarify what you mean by that?
> Assets are always resolved per target framework and the 2 projects you're listing don't have the same target frameworks.

It seemed to me that the packages for both projects are only resolved once and when the transitive one was first it did not scan for any build dir in the packages. Don't take my word for it however, I don't know the code very well and just debugged to see if we were doing dumb things in our packages.

--- a/src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs
+++ b/src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs
@@ -78,6 +78,14 @@ namespace NuGet.Frameworks
 
                 combiner.AddObject(Comparer.GetHashCode(this));
 
+                // Include DualCompatibilityFramework's secondary framework in the hash
+                // to distinguish ATF(DCF(net6.0, native)) from ATF(net6.0).
+                if (RootFramework is DualCompatibilityFramework dualCompatibilityFramework)
+                {
+                    combiner.AddStringIgnoreCase(nameof(DualCompatibilityFramework));
+                    combiner.AddObject(Comparer.GetHashCode(dualCompatibilityFramework.SecondaryFramework));
+                }
+
                 foreach (var each in Fallback)
                 {
                     combiner.AddObject(Comparer.GetHashCode(each));
@@ -101,8 +109,26 @@ namespace NuGet.Frameworks
                 return true;
             }
 
-            return NuGetFramework.Comparer.Equals(this, other)
-                && Fallback.SequenceEqual(other.Fallback);
+            if (!NuGetFramework.Comparer.Equals(this, other))
+            {
+                return false;
+            }
+
+            // Check DualCompatibilityFramework secondary framework equality to distinguish
+            // ATF(DCF(net6.0, native)) from ATF(net6.0).
+            var thisDcf = RootFramework as DualCompatibilityFramework;
+            var otherDcf = other.RootFramework as DualCompatibilityFramework;
+            if ((thisDcf == null) != (otherDcf == null))
+            {
+                return false;
+            }
+
+            if (thisDcf != null && !Comparer.Equals(thisDcf.SecondaryFramework, otherDcf!.SecondaryFramework))
+            {
+                return false;
+            }
+
+            return Fallback.SequenceEqual(other.Fallback);
         }
     }
 }

--- a/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs
@@ -4711,7 +4711,6 @@ namespace NuGet.Commands.FuncTest
             }
         }
 
-        // https://github.com/NuGet/Home/issues/13326
         // When a managed project (B) references a C++/CLI project (A) that has a package
         // with build/native/ targets, restoring B first causes A to miss the build/native/ targets
         // due to a LockFileBuilderCache collision. When both projects use AssetTargetFallback with

--- a/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs
@@ -4711,6 +4711,129 @@ namespace NuGet.Commands.FuncTest
             }
         }
 
+        // https://github.com/NuGet/Home/issues/13326
+        // When a managed project (B) references a C++/CLI project (A) that has a package
+        // with build/native/ targets, restoring B first causes A to miss the build/native/ targets
+        // due to a LockFileBuilderCache collision. When both projects use AssetTargetFallback with
+        // the same imports, NuGetFrameworkFullComparer treats DualCompatibilityFramework(net6.0, native)
+        // as equal to plain net6.0, causing the criteria cache to return wrong selection criteria.
+        [Fact]
+        public async Task RestoreCommand_CppCliWithNativeBuildTargets_SharedCache_RestoreOrderDoesNotAffectBuildAssets()
+        {
+            using var pathContext = new SimpleTestPathContext();
+
+            // Package structure:
+            // MyPackage: build/native/MyPackage.targets (only matched by native framework criteria)
+            // MyPackage depends on MyPackage-binaries
+            // MyPackage-binaries: buildTransitive/MyPackage-binaries.targets
+            var binariesPackage = new SimpleTestPackageContext("MyPackage-binaries", "1.0.0");
+            binariesPackage.AddFile("buildTransitive/MyPackage-binaries.targets");
+
+            var myPackage = new SimpleTestPackageContext("MyPackage", "1.0.0");
+            myPackage.AddFile("build/native/MyPackage.targets");
+            myPackage.Dependencies.Add(binariesPackage);
+
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                pathContext.PackageSource,
+                PackageSaveMode.Defaultv3,
+                myPackage,
+                binariesPackage);
+
+            // Project A: C++/CLI (net6.0-windows7.0 + secondaryFramework=native + ATF)
+            // Direct PackageReference to MyPackage
+            var projectAJson = @"
+                {
+                    ""frameworks"": {
+                        ""net6.0-windows7.0"": {
+                            ""targetAlias"" : ""net6.0-windows7.0"",
+                            ""assetTargetFallback"" : true,
+                            ""imports"" : [
+                                ""net461"",
+                                ""net472""
+                            ],
+                            ""secondaryFramework"" : ""native"",
+                            ""dependencies"": {
+                                ""MyPackage"": ""1.0.0""
+                            }
+                        }
+                    }
+                }";
+
+            // Project B: managed net6.0-windows7.0 with same ATF imports, references Project A
+            var projectBJson = @"
+                {
+                    ""frameworks"": {
+                        ""net6.0-windows7.0"": {
+                            ""targetAlias"" : ""net6.0-windows7.0"",
+                            ""assetTargetFallback"" : true,
+                            ""imports"" : [
+                                ""net461"",
+                                ""net472""
+                            ],
+                            ""dependencies"": {
+                            }
+                        }
+                    }
+                }";
+
+            var logger = new TestLogger();
+
+            var projectA = ProjectTestHelpers.GetPackageSpecWithProjectNameAndSpec(
+                "ProjectA", pathContext.SolutionRoot, projectAJson);
+            var projectB = ProjectTestHelpers.GetPackageSpecWithProjectNameAndSpec(
+                "ProjectB", pathContext.SolutionRoot, projectBJson);
+
+            // B references A (default PrivateAssets = contentfiles;analyzers;build)
+            projectB = projectB.WithTestProjectReference(projectA);
+            CreateFakeProjectFile(projectA);
+
+            // Use DependencyGraphSpecRequestProvider which naturally shares a LockFileBuilderCache
+            // across all projects in the solution restore
+            var dgSpec = ProjectTestHelpers.GetDGSpecForAllProjects(projectB, projectA);
+
+            var restoreContext = new RestoreArgs()
+            {
+                CacheContext = new SourceCacheContext(),
+                Log = logger,
+                GlobalPackagesFolder = pathContext.UserPackagesFolder,
+                Sources = new List<string>() { pathContext.PackageSource },
+            };
+
+            var dgProvider = new DependencyGraphSpecRequestProvider(
+                new RestoreCommandProvidersCache(),
+                dgSpec);
+
+            IReadOnlyList<RestoreSummaryRequest> requests = await dgProvider.CreateRequests(restoreContext);
+
+            // Restore B first, then A (the problematic order)
+            var requestB = requests.Single(r => r.Request.Project.Name.Equals("ProjectB", StringComparison.OrdinalIgnoreCase));
+            var requestA = requests.Single(r => r.Request.Project.Name.Equals("ProjectA", StringComparison.OrdinalIgnoreCase));
+
+            var resultB = await new RestoreCommand(requestB.Request).ExecuteAsync();
+            resultB.Success.Should().BeTrue();
+
+            var resultA = await new RestoreCommand(requestA.Request).ExecuteAsync();
+            resultA.Success.Should().BeTrue();
+
+            // Assert: A's lock file should include build/native/MyPackage.targets
+            resultA.LockFile.Targets.Should().HaveCount(1);
+            var targetA = resultA.LockFile.Targets[0];
+
+            var myPackageLib = targetA.Libraries.First(l => l.Name.Equals("MyPackage", StringComparison.OrdinalIgnoreCase));
+            myPackageLib.Build.Should().ContainSingle()
+                .Which.Path.Should().Be("build/native/MyPackage.targets");
+
+            var binariesLib = targetA.Libraries.First(l => l.Name.Equals("MyPackage-binaries", StringComparison.OrdinalIgnoreCase));
+            binariesLib.Build.Should().ContainSingle()
+                .Which.Path.Should().Be("buildTransitive/MyPackage-binaries.targets");
+
+            // Assert: B should not get build assets (transitive through P2P with PrivateAssets=build)
+            resultB.LockFile.Targets.Should().HaveCount(1);
+            var targetB = resultB.LockFile.Targets[0];
+            var myPackageLibB = targetB.Libraries.First(l => l.Name.Equals("MyPackage", StringComparison.OrdinalIgnoreCase));
+            myPackageLibB.Build.Should().BeEmpty();
+        }
+
         private static JObject BasicConfigWithNet46
         {
             get

--- a/test/NuGet.Core.Tests/NuGet.Frameworks.Test/AssetTargetFallbackTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Frameworks.Test/AssetTargetFallbackTests.cs
@@ -64,5 +64,52 @@ namespace NuGet.Frameworks.Test
             var assetTargetFallback = new AssetTargetFallbackFramework(nugetFramework, fallbackFrameworks: SampleFrameworkList);
             Assert.False(assetTargetFallback.Equals((object)nugetFramework));
         }
+
+        [Fact]
+        public void Equals_WithDualCompatibilityFramework_DifferentFromPlainFramework()
+        {
+            var net6Win = NuGetFramework.Parse("net6.0-windows7.0");
+            var native = NuGetFramework.Parse("native");
+            var fallbacks = new NuGetFramework[] { NuGetFramework.Parse("net461") };
+
+            var atfPlain = new AssetTargetFallbackFramework(net6Win, fallbacks);
+            var atfDcf = new AssetTargetFallbackFramework(
+                new DualCompatibilityFramework(net6Win, native), fallbacks);
+
+            atfPlain.Equals(atfDcf).Should().BeFalse(
+                because: "ATF wrapping DualCompatibilityFramework(net6.0, native) must differ from ATF wrapping plain net6.0");
+            atfDcf.Equals(atfPlain).Should().BeFalse();
+        }
+
+        [Fact]
+        public void GetHashCode_WithDualCompatibilityFramework_DifferentFromPlainFramework()
+        {
+            var net6Win = NuGetFramework.Parse("net6.0-windows7.0");
+            var native = NuGetFramework.Parse("native");
+            var fallbacks = new NuGetFramework[] { NuGetFramework.Parse("net461") };
+
+            var atfPlain = new AssetTargetFallbackFramework(net6Win, fallbacks);
+            var atfDcf = new AssetTargetFallbackFramework(
+                new DualCompatibilityFramework(net6Win, native), fallbacks);
+
+            atfPlain.GetHashCode().Should().NotBe(atfDcf.GetHashCode(),
+                because: "hash codes must differ to avoid cache collisions in LockFileBuilderCache");
+        }
+
+        [Fact]
+        public void Equals_TwoDualCompatibilityFrameworks_WithSameSecondary_AreEqual()
+        {
+            var net6Win = NuGetFramework.Parse("net6.0-windows7.0");
+            var native = NuGetFramework.Parse("native");
+            var fallbacks = new NuGetFramework[] { NuGetFramework.Parse("net461") };
+
+            var atf1 = new AssetTargetFallbackFramework(
+                new DualCompatibilityFramework(net6Win, native), fallbacks);
+            var atf2 = new AssetTargetFallbackFramework(
+                new DualCompatibilityFramework(net6Win, native), fallbacks);
+
+            atf1.Equals(atf2).Should().BeTrue();
+            atf1.GetHashCode().Should().Be(atf2.GetHashCode());
+        }
     }
 }

